designer/server/src/main/scala/pl/touk/nussknacker/ui/process/newdeployment/DeploymentService.scala

```diff
@@ -0,0 +1,145 @@
+    // TODO: we shouldn't call legacy service, instead we should new service from the legacy one
+    legacyDeploymentService: LegacyDeploymentService,
+    dbioRunner: DBIOActionRunner,
+    clock: Clock
+)(implicit ec: ExecutionContext) {
+
+  def processCommand(
+      command: DeploymentCommand
+  ): Future[Either[RunDeploymentError, Unit]] = {
+    command match {
+      case command: RunDeploymentCommand =>
+        runDeployment(command)
+    }
+  }
+
+  private def runDeployment(
+      command: RunDeploymentCommand
+  ): Future[Either[RunDeploymentError, Unit]] =
+    (for {
+      scenarioMetadata <- getScenarioMetadata(command)
+      _ <- EitherT.cond[Future](
+        command.user.can(scenarioMetadata.processCategory, Permission.Deploy),
+        (),
+        NoPermissionError
+      )
+      _         <- saveDeployment(command, scenarioMetadata)
+      runResult <- invokeLegacyRunDeploymentLogic(command, scenarioMetadata)
+    } yield runResult).value
+
+  private def getScenarioMetadata(command: RunDeploymentCommand) =
+    EitherT[Future, RunDeploymentError, ProcessEntityData](
+      dbioRunner.run(
+        scenarioMetadataRepository
+          .getScenarioMetadata(command.scenarioName)
+          .map(_.map(Right(_)).getOrElse(Left(ScenarioNotFoundError(command.scenarioName))))
+      )
+    )
+
+  private def saveDeployment(command: RunDeploymentCommand, scenarioMetadata: ProcessEntityData) =
+    EitherT(
+      dbioRunner.run(
+        deploymentRepository.saveDeployment(
+          DeploymentEntityData(command.id, scenarioMetadata.id, Timestamp.from(clock.instant()), command.user.id)
+        )
+      )
+    ).leftMap { e => ConflictingDeploymentIdError(e.id) }
+
+  private def invokeLegacyRunDeploymentLogic(command: RunDeploymentCommand, scenarioMetadata: ProcessEntityData) = {
+    EitherT(
+      // TODO: Currently it doesn't use our deploymentId. Instead, it uses action id
+      legacyDeploymentService
+        .processCommand(
+          LegacyRunDeploymentCommand(
+            CommonCommandData(
+              ProcessIdWithName(scenarioMetadata.id, command.scenarioName),
+              command.comment,
+              command.user
+            ),
+            savepointPath = None,
+            nodesDeploymentData = command.nodesDeploymentData,
+          )
+        )
+        .transform { result =>
+          result
+            .map[Either[RunDeploymentError, Unit]](_ => Right(()))
+            .recover { case CommentValidationError(msg) =>
+              Left(NewCommentValidationError(msg))
+            }
+        }
+    )
+  }
+
+  def getDeploymentStatus(
+      id: DeploymentId
+  )(implicit loggedUser: LoggedUser): Future[Either[GetDeploymentStatusError, StatusName]] =
+    (for {
+      deploymentWithScenarioMetadata <- EitherT(
+        dbioRunner.run(deploymentRepository.getDeploymentById(id))
+      )
+      DeploymentWithScenarioMetadata(_, scenarioMetadata) = deploymentWithScenarioMetadata
```

Comment:
I'm not sure if this is needed. We can just access the field. Do we really need to unapply this?

Reply:
I don't like long expressions like deploymentWithScenarioMetadata.scenarioMetadata but if you have some allergy to destructing assignment, I can change it :)

Reply:
I general, no I don't. But I see a value in readability when the public method looks like this:
```
def businessMethod(...) {
   for {
     doSth
     checkSth
     writeSth
   } yield result
}
```

And when there is in the middle some form of distraction (I mean, sth technical), the readability is worse. I want to focus in this place on the algorithm (business logic).